validator/client/src/test/java/tech/pegasys/artemis/validator/client/DutySchedulerTest.java

```diff
@@ -0,0 +1,191 @@
+
+    // Should retry request
+    verify(validatorApiChannel, times(2)).getDuties(UnsignedLong.ONE, VALIDATOR_KEYS);
+
+    // And not have any more retries scheduled
+    assertThat(asyncRunner.hasDelayedActions()).isFalse();
+  }
+
+  @Test
+  public void shouldTimeoutRequestsForDuties() {
+    fail("Not implemented");
+  }
+
+  @Test
+  public void shouldScheduleBlockProposalDuty() {
+    final UnsignedLong blockProposerSlot = UnsignedLong.valueOf(5);
+    final ValidatorDuties validator1Duties =
+        new ValidatorDuties(
+            VALIDATOR1_KEY, 5, UnsignedLong.ZERO, 3, Optional.empty(), List.of(blockProposerSlot));
+    when(validatorApiChannel.getDuties(eq(UnsignedLong.ONE), any()))
+        .thenReturn(SafeFuture.completedFuture(List.of(validator1Duties)));
+
+    final Duty blockCreationDuty = mock(Duty.class);
+    when(dutyFactory.createBlockProductionDuty(validator1, blockProposerSlot))
+        .thenReturn(blockCreationDuty);
+
+    // Load duties
+    validatorClient.onEpoch(UnsignedLong.ONE);
+
+    // Execute
+    validatorClient.onBlockProductionDue(blockProposerSlot);
+    verify(blockCreationDuty).performDuty();
+  }
+
+  @Test
+  public void shouldNotPerformDutiesForSameSlotTwice() {
+    final UnsignedLong blockProposerSlot = UnsignedLong.valueOf(5);
+    final ValidatorDuties validator1Duties =
+        new ValidatorDuties(
+            VALIDATOR1_KEY, 5, UnsignedLong.ZERO, 3, Optional.empty(), List.of(blockProposerSlot));
+    when(validatorApiChannel.getDuties(eq(UnsignedLong.ONE), any()))
+        .thenReturn(SafeFuture.completedFuture(List.of(validator1Duties)));
+
+    final Duty blockCreationDuty = mock(Duty.class);
+    when(dutyFactory.createBlockProductionDuty(validator1, blockProposerSlot))
+        .thenReturn(blockCreationDuty);
+
+    // Load duties
+    validatorClient.onEpoch(UnsignedLong.ONE);
+
+    // Execute
+    validatorClient.onBlockProductionDue(blockProposerSlot);
+    verify(blockCreationDuty).performDuty();
+
+    // Somehow we triggered the same slot again.
+    validatorClient.onBlockProductionDue(blockProposerSlot);
+    // But shouldn't produce another block and get ourselves slashed.
+    verifyNoMoreInteractions(blockCreationDuty);
+  }
+
+  @Test
+  public void shouldScheduleAttestationDuties() {
+    final UnsignedLong attestationSlot = UnsignedLong.valueOf(5);
+    final ValidatorDuties validator1Duties =
+        new ValidatorDuties(
+            VALIDATOR1_KEY, 5, UnsignedLong.ZERO, 3, Optional.of(attestationSlot), emptyList());
+    final ValidatorDuties validator2Duties =
+        new ValidatorDuties(
+            VALIDATOR2_KEY, 6, UnsignedLong.ZERO, 4, Optional.of(attestationSlot), emptyList());
+    when(validatorApiChannel.getDuties(eq(UnsignedLong.ONE), any()))
+        .thenReturn(SafeFuture.completedFuture(List.of(validator1Duties, validator2Duties)));
+
+    final AttestationProductionDuty attestationDuty = mock(AttestationProductionDuty.class);
```

Comment:
maybe move this up top with declarations so the when's can be together? makes it easier to follow...